### PR TITLE
TW-858: Jump to unread msg when click on notification

### PR DESCRIPTION
**Pull request by nqhhdev**
### Ticket:
- #858 

#### MOBILE

https://github.com/linagora/twake-on-matrix/assets/99852347/f1c7136c-0355-4e99-a48c-7a74d51ea0a2

#### WEB

https://github.com/linagora/twake-on-matrix/assets/99852347/674fd96b-3d99-4560-8fd6-8e62aecf8af4

https://github.com/linagora/twake-on-matrix/assets/99852347/eb0f3ba3-2a74-4453-bdf2-612acab036ce

### CODE REVIEW

File: lib/pages/chat/events/message/message.dart [hunk 1]
```diff
@@ -456,6 +458,11 @@ class Message extends StatelessWidget {
                 StickyTimestampWidget(
                   content: event.originServerTs.relativeTime(context),
                 ),
+              if (unreadEvent != null && unreadEvent?.eventId == event.eventId)
+                Text(
+                  "Unread message",
+                  style: Theme.of(context).textTheme.bodySmall,
+                ),
```
Comment (hoangdat) [inline lib/pages/chat/events/message/message.dart:465]: We should make it more stand out

File: lib/pages/chat/chat_event_list.dart [hunk 2]
```diff
@@ -106,36 +106,45 @@ class ChatEventList extends StatelessWidget {
...
-                        outerContext: context,
-                        onMention: () => controller.sendController.text +=
-                            '${event.senderFromMemoryOrFallback.mention} ',
+                ? VisibilityDetector(
+                    key: ValueKey(event.eventId),
+                    onVisibilityChanged: (visibilityInfo) =>
+                        controller.handleMessageVisibilityChanged(
+                      event,
+                      visibilityInfo.visibleFraction > 0.1,
+                    ),
```
Comment (hoangdat) [inline lib/pages/chat/chat_event_list.dart:115]: Please write an ADR about the way we set read marker, what we can improve....

File: lib/pages/chat/chat.dart [hunk 3]
```diff
@@ -117,6 +117,8 @@ class ChatController extends State<Chat>
 
   PinnedEventsController pinnedEventsController = PinnedEventsController();
 
+  static const waitForJumpToFirstUnreadEvent = Duration(seconds: 1);
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:120]: what happen if it take more time than that?

File: lib/pages/chat/chat.dart [hunk 4]
```diff
@@ -322,6 +327,67 @@ class ChatController extends State<Chat>
     }
   }
 
+  List<Event?> getEventsUnread() {
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:330]:
```suggestion
  List<Event?> getUnreadEvents() {
```

File: lib/pages/chat/chat.dart [hunk 5]
```diff
@@ -322,6 +327,67 @@ class ChatController extends State<Chat>
...
+
+    return [];
+  }
+
+  Event? getFirstUnreadEvent() {
+    final unreadEvents = getEventsUnread();
+    if (unreadEvents.isEmpty) return null;
+
+    return unreadEvents.last;
+  }
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:357]: `first` but return `last`?

File: lib/pages/chat/chat.dart [hunk 6]
```diff
@@ -322,6 +327,67 @@ class ChatController extends State<Chat>
...
+
+  void handleMessageVisibilityChanged(Event event, bool visible) {
+    if (_isJumpingToLastReadEvent) return;
+    Logs().d(
+      'Chat::handleMessageVisibilityChanged() - isVisible $visible',
+    );
+    final unreadEvents = getEventsUnread();
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:364]: IMO, it is too expensive behavior.

File: lib/pages/chat/chat.dart [hunk 7]
```diff
@@ -322,6 +327,67 @@ class ChatController extends State<Chat>
     }
   }
 
+  List<Event?> getEventsUnread() {
+    final fullyRead = room!.fullyRead;
+    final lastIndexReadEvent = timeline?.events.indexWhere(
+      (event) => fullyRead.contains(event.eventId),
```
Comment (drminh2807) [inline lib/pages/chat/chat.dart:333]:
```suggestion
      (event) => fullyRead == event.eventId,
```

File: lib/pages/chat/chat.dart [hunk 8]
```diff
@@ -363,16 +431,20 @@ class ChatController extends State<Chat>
     context.pop();
   }
 
+  void _jumpToLastMessageFullyRead() async {
+    await getTimeline();
+    if (room == null || room!.notificationCount <= 0) return;
+    final lastReadEvent = room!.fullyRead;
+    await Future.delayed(waitForJumpToFirstUnreadEvent);
+    _isJumpingToLastReadEvent = true;
+    scrollToEventId(lastReadEvent);
```
Comment (drminh2807) [inline lib/pages/chat/chat.dart:440]: scrollToEventId is a future, should wait for it, instead `waitForJumpToFirstUnreadEvent`

File: lib/pages/chat/chat.dart [hunk 9]
```diff
@@ -322,6 +327,67 @@ class ChatController extends State<Chat>
...
   }
 
+  List<Event?> getEventsUnread() {
+    final fullyRead = room!.fullyRead;
+    final lastIndexReadEvent = timeline?.events.indexWhere(
+      (event) => fullyRead.contains(event.eventId),
+    );
+    if (lastIndexReadEvent != -1 &&
+        lastIndexReadEvent != 0 &&
+        lastIndexReadEvent != null) {
```
Comment (sherlockvn) [inline lib/pages/chat/chat.dart:337]:
```suggestion
    if (lastIndexReadEvent != null &&
    lastIndexReadEvent != -1 &&
        lastIndexReadEvent != 0
        ) {
```

File: lib/pages/chat/events/message/message.dart [hunk 10]
```diff
@@ -456,6 +458,11 @@ class Message extends StatelessWidget {
                 StickyTimestampWidget(
                   content: event.originServerTs.relativeTime(context),
                 ),
+              if (unreadEvent != null && unreadEvent?.eventId == event.eventId)
+                Text(
+                  "Unread message",
```
Comment (sherlockvn) [inline lib/pages/chat/events/message/message.dart:463]: l10n

File: lib/pages/chat/chat.dart [hunk 11]
```diff
@@ -322,6 +327,67 @@ class ChatController extends State<Chat>
     }
   }
 
+  List<Event?> getEventsUnread() {
+    final fullyRead = room!.fullyRead;
+    final lastIndexReadEvent = timeline?.events.indexWhere(
+      (event) => fullyRead.contains(event.eventId),
+    );
+    if (lastIndexReadEvent != -1 &&
+        lastIndexReadEvent != 0 &&
```
Comment (sherlockvn) [inline lib/pages/chat/chat.dart:336]: lastIndexReadEvent != null && lastIndexReadEvent > 0 &&

File: lib/pages/chat/chat.dart (hunk 3 above)
Comment (nqhhdev) [reply lib/pages/chat/chat.dart:120]: I think no need it.

File: lib/pages/chat/chat.dart [hunk 12]
```diff
@@ -322,6 +327,67 @@ class ChatController extends State<Chat>
...
+      final unreadEvents = afterFullyRead?.map((fullyReadEvent) {
+        if (fullyReadEvent.senderId != client.userID) {
+          return fullyReadEvent;
+        }
+      }).toList();
+      if (unreadEvents == null || unreadEvents.isEmpty) return [];
+      Logs().d(
+        "Chat::getFirstUnreadEvent(): Last unread event ${unreadEvents.last?.eventId}",
+      );
+      _unreadEvents = unreadEvents.toSet();
```
Comment (sherlockvn) [inline lib/pages/chat/chat.dart:346]: i think no need `toSet`, because it's always different to each other

File: assets/l10n/intl_en.arb [hunk 13]
```diff
@@ -2927,5 +2927,6 @@
   "contactsCount": "Contacts ({count})",
   "linagoraContactsCount": "Linagora contacts ({count})",
   "fetchingPhonebookContacts": "Fetching contacts from device...({progress}% completed)",
-  "activeNow": "Active now"
+  "activeNow": "Active now",
+  "unreadMessage": "Unread message"
```
Comment (sherlockvn) [inline assets/l10n/intl_en.arb:2931]:
```suggestion
  "unreadMessage": "Unread messages"
```

File: lib/pages/chat/chat.dart [hunk 14]
```diff
@@ -322,6 +327,67 @@ class ChatController extends State<Chat>
...
+    final fullyRead = room!.fullyRead;
+    final lastIndexReadEvent = timeline?.events.indexWhere(
+      (event) => fullyRead == event.eventId,
+    );
+    if (lastIndexReadEvent != null && lastIndexReadEvent > 0) {
+      final afterFullyRead = timeline?.events.getRange(0, lastIndexReadEvent);
+      final unreadEvents = afterFullyRead?.map((fullyReadEvent) {
+        if (fullyReadEvent.senderId != client.userID) {
+          return fullyReadEvent;
+        }
```
Comment (imGok) [inline lib/pages/chat/chat.dart:340]: why not use `where` ?
```suggestion
      final unreadEvents = afterFullyRead?.where(fullyReadEvent => fullyReadEvent.senderId != client.userID);
```

File: lib/pages/chat/chat.dart (hunk 8 above)
Comment (nqhhdev) [reply lib/pages/chat/chat.dart:443]: It's a void func, not future

File: lib/pages/chat/chat_event_list.dart [hunk 15]
```diff
@@ -112,36 +113,45 @@ class ChatEventList extends StatelessWidget {
...
-                          user: event.senderFromMemoryOrFallback,
-                          outerContext: context,
-                          onMention: () => controller.sendController.text +=
-                              '${event.senderFromMemoryOrFallback.mention} ',
+                  ? VisibilityDetector(
+                      key: ValueKey(event.eventId),
+                      onVisibilityChanged: (visibilityInfo) =>
+                          controller.handleMessageVisibilityChanged(
+                        event,
+                        visibilityInfo.visibleFraction > 0.1,
```
Comment (sherlockvn) [inline lib/pages/chat/chat_event_list.dart:121]: const

**Pull request closed without merging**